### PR TITLE
fix: Add Utils methods for slotIndex and splitting outputId into parts

### LINKED ISSUE #1987: Revert IdWithSlotIndex changes and add Util for computing slotIndex
## Description

The solution added in this [PR](https://github.com/iotaledger/iota-sdk/pull/1962) doesn't work for Explorer, as `IdWithSlotIndex` being a subclass of build-in type `String` doesn't import properly.

We agreed in a [discussion in slack](https://iotafoundation.slack.com/archives/C041M474AKE/p1707817612797929) to revert those changes and simply add `Utils` to compute a slotIndex form block/transaction id. Something like:
```
static slotIndex(id: BlockId | TransactionId): number {
        const numberString = this.id.slice(-8);
        const chunks = [];
        for (
            let i = 0, charsLength = numberString.length;
            i < charsLength;
            i += 2
        ) {
            chunks.push(numberString.substring(i, i + 2));
        }
        const separated = chunks.map((n) => parseInt(n, 16));
        const buf = Uint8Array.from(separated).buffer;
        const view = new DataView(buf);
        return view.getUint32(0, true);
}
```

but with a call into rust bindings for the computation.

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by msarcev**
# Description of change

* Reverts changes of https://github.com/iotaledger/iota-sdk/pull/1962
* Adds new Utils methods

fixes https://github.com/iotaledger/iota-sdk/issues/1987

## Links to any relevant issues

fixes https://github.com/iotaledger/iota-sdk/issues/1987

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: bindings/nodejs/lib/utils/utils.ts [hunk 1]
```diff
@@ -411,6 +448,29 @@ export class Utils {
         });
     }
 
+    /**
+     * Computes a slotIndex from a block or transaction Id.
+     * @param id The block or transaction Id.
+     * @returns The slotIndex.
+     */
+    static computeSlotIndex(id: BlockId | TransactionId): SlotIndex {
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/utils/utils.ts]:
```suggestion
    static computeSlotIndex(id: BlockId | SlotCommitmentId | TransactionId): SlotIndex {
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
CI failing

File: bindings/nodejs/tests/types/block.spec.ts [hunk 2]
```diff
@@ -16,23 +16,23 @@ describe('Block tests', () => {
         const block = parseBlock(basic_block_tagged_data_payload_json.block);
         expect(block).toBeInstanceOf(Block);
         const params: ProtocolParameters = JSON.parse(JSON.stringify(protocol_parameters_json.params));
-        const expected_id = basic_block_tagged_data_payload_json.id as BlockId;
+        const expected_id = basic_block_tagged_data_payload_json.id as unknown as BlockId;
```
Comment (thibault-martinez) [inline bindings/nodejs/tests/types/block.spec.ts]: As a non-TS person, this looks extremely weird

File: bindings/nodejs/tests/types/block.spec.ts [hunk 3]
```diff
@@ -11,28 +11,27 @@ import * as protocol_parameters_json from '../../../../sdk/tests/types/fixtures/
...
+        const expected_id: BlockId = basic_block_transaction_payload_json.id;
         expect(Block.id(block, params)).toEqual(expected_id);
     });
 
     it('compares validation block from a fixture', async () => {
         const block = parseBlock(validation_block_json.block);
         expect(block).toBeInstanceOf(Block);
         const params: ProtocolParameters = JSON.parse(JSON.stringify(protocol_parameters_json.params));
-        const expected_id = validation_block_json.id as BlockId;
+        const expected_id = validation_block_json.id;
```
Comment (thibault-martinez) [inline bindings/nodejs/tests/types/block.spec.ts]: Curiosity, why is the explicit type not needed here?

File: bindings/nodejs/lib/utils/utils.ts [hunk 4]
```diff
@@ -411,6 +448,31 @@ export class Utils {
         });
     }
 
+    /**
+     * Computes a slotIndex from a block or transaction Id.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/utils/utils.ts]: Or slot commitment Id?

File: bindings/nodejs/lib/utils/utils.ts [hunk 5]
```diff
@@ -411,6 +448,31 @@ export class Utils {
...
+     * @param id The block or transaction Id.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/utils/utils.ts]: same

File: bindings/nodejs/tests/types/block.spec.ts (hunk 3 above)
Comment (msarcev) [reply bindings/nodejs/tests/types/block.spec.ts]: It's cause `expected_id` is a `string`, and `BlockId` is just an "alias" type for `string`, so it's the same (the next link that asserts doesn't complain cause the type matches). But i'll add it for consistency.

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**